samples/tiled-convolution/tiled-convolution.cpp

```diff
@@ -0,0 +1,310 @@
+  // mask(filter) SYCL buffer
+  auto fil_buff = cl::sycl::buffer<data_t, 2>(
+      filter.data(), cl::sycl::range<2>(fil_size.m, fil_size.n));
+  // output SYCL buffer
+  auto out_buff = cl::sycl::buffer<data_t, 2>(
+      cl::sycl::range<2>(total_buffer.m, total_buffer.n));
+  double total_submission_time = 0;
+  double total_execution_time = 0;
+  double avarage_submission_time = 0;
+  double avarage_execution_time = 0;
+  double avarage_application_execution_time = 0;
+  double total_application_execution_time = 0;
+  index_t host_offset_m = 0;
+  // launching tiled-based kernel via two nested for-loop
+  for (index_t m = 0; m < num_host_tile_m; m++) {
+    index_t host_offset_n = 0;
+    for (index_t n = 0; n < num_host_tile_n; n++) {
+      index_t range_src_m, offset_src_m;
+      index_t range_src_n, offset_src_n;
+      // calculating the halo for first dimension of the tile
+      compute_index(total_buffer.m, mat_size.m, fil_size.m, host_offset_m,
+                    range_src_m, offset_src_m);
+      // calculating the halo for the second dimension of the tile
+      compute_index(total_buffer.n, mat_size.n, fil_size.n, host_offset_n,
+                    range_src_n, offset_src_n);
+      auto start = std::chrono::system_clock::now();
+      // events[n + m * num_host_tile_n] =
+      auto event = sycl_queue.submit([&](cl::sycl::handler &cgh) {
+        // filter
+        auto fil_acc = fil_buff.get_access<read_t, global_buffer_t>(cgh);
+        // input
+        auto in_acc = in_buff.get_access<read_t, global_buffer_t>(
+            cgh, cl::sycl::range<2>(range_src_m, range_src_n),
+            cl::sycl::id<2>(offset_src_m, offset_src_n));
+        // output
+        auto out_acc = out_buff.get_access<write_t, global_buffer_t>(
+            cgh, cl::sycl::range<2>(mat_size.m, mat_size.n),
+            cl::sycl::id<2>(host_offset_m, host_offset_n));
+        // global size m
+        auto global_size_m = round_up(mat_size.m, ocl_config_t::local_size_m);
+        // global size n
+        auto global_size_n = round_up(mat_size.n, ocl_config_t::local_size_n);
+        // constructing the kernel
+        cgh.parallel_for(cl::sycl::nd_range<2>(
+                             cl::sycl::range<2>(global_size_m, global_size_n),
+                             cl::sycl::range<2>(ocl_config_t::local_size_m,
+                                                ocl_config_t::local_size_n),
+                             cl::sycl::id<2>(host_offset_m, host_offset_n)),
+                         conv<read_accessor_t, write_accessor_t, index_t,
+                              matrix_size_t<index_t>>(fil_acc, in_acc, out_acc,
+                                                      total_buffer, fil_size));
+      });
+// this part is used to profile each tiled kernel
+#ifdef PROFILE_SYCL
+      auto i = n + m * num_host_tile_n;
+      event.wait();
+      auto end = std::chrono::system_clock::now();
```

Comment:
The first iteration take longer than the others, because it will include the module compilation time. You could use the program class to avoid this (though some implementations might still have compiling to do, it should still help).